line_profiler/_line_profiler.pyx

```diff
@@ -1462,33 +1460,28 @@ cdef inline inner_trace_callback(
             time = hpTimer()
             has_time = True
         ident = PyThread_get_thread_ident()
-        last_map = &(prof._c_last_time[ident])
         # deref() is Cython's version of the -> accessor in C++. if we don't use deref then
         # Cython thinks that when we index last_map,
         # we want pointer indexing (which is not the case)
-        if deref(last_map).count(block_hash):
-            old = deref(last_map)[block_hash]
+        last_map = &(prof._c_last_time[ident])
+        if last_contains(last_map, block_hash):
+            # Read once; no extra insertion
+            old = last_get_value(last_map, block_hash)
             line_entries = &(prof._c_code_map[code_hash])
-            key = old.f_lineno
-            if not deref(line_entries).count(key):
-                deref(line_entries)[key] = LineTime(code_hash, key, 0, 0)
-            entry = &(deref(line_entries)[key])
-            # Note: explicitly `deref()`-ing here causes the new values
-            # to be assigned to a temp var;
-            # meanwhile, directly dot-accessing a pointer causes Cython
-            # to correctly write `ptr->attr = (ptr->attr + incr)`
+            entry = line_ensure_entry(line_entries, old.f_lineno, code_hash)
             entry.nhits += 1
             entry.total_time += time - old.time
+            have_last = True
+        else:
+            have_last = False
+
         if is_line_event:
-            # Get the time again. This way, we don't record much time
-            # wasted in this function.
-            deref(last_map)[block_hash] = LastTime(lineno, hpTimer())
-        elif deref(last_map).count(block_hash):
-            # We are returning from a function, not executing a line.
-            # Delete the last_time record. It may have already been
-            # deleted if we are profiling a generator that is being
-            # pumped past its end.
-            deref(last_map).erase(deref(last_map).find(block_hash))
+            # Record fresh timestamp for this block
+            last_set_now(last_map, block_hash, lineno)
+        elif have_last:
+            # Returning from a function: cleanup
+            last_erase_if_present(last_map, block_hash)
```

Comment:
Can we keep the comment that previously existed here?

Reply:
yes, if you could make a final pass of cherry picking my suggestions in this branch onto your PR, that would help to get it tested on CI.

Reply:
Sure!